volatility3/framework/plugins/windows/devicetree.py

```diff
@@ -0,0 +1,145 @@
+    0x0000000e : "FILE_DEVICE_MIDI_OUT",
+    0x0000002b : "FILE_DEVICE_MODEM",
+    0x0000000f : "FILE_DEVICE_MOUSE",
+    0x00000010 : "FILE_DEVICE_MULTI_UNC_PROVIDER",
+    0x00000011 : "FILE_DEVICE_NAMED_PIPE",
+    0x00000012 : "FILE_DEVICE_NETWORK",
+    0x00000013 : "FILE_DEVICE_NETWORK_BROWSER",
+    0x00000014 : "FILE_DEVICE_NETWORK_FILE_SYSTEM",
+    0x00000028 : "FILE_DEVICE_NETWORK_REDIRECTOR",
+    0x00000015 : "FILE_DEVICE_NULL",
+    0x00000016 : "FILE_DEVICE_PARALLEL_PORT",
+    0x00000017 : "FILE_DEVICE_PHYSICAL_NETCARD",
+    0x00000018 : "FILE_DEVICE_PRINTER",
+    0x00000019 : "FILE_DEVICE_SCANNER",
+    0x0000001c : "FILE_DEVICE_SCREEN",
+    0x00000037 : "FILE_DEVICE_SERENUM",
+    0x0000001a : "FILE_DEVICE_SERIAL_MOUSE_PORT",
+    0x0000001b : "FILE_DEVICE_SERIAL_PORT",
+    0x00000031 : "FILE_DEVICE_SMARTCARD",
+    0x0000002e : "FILE_DEVICE_SMB",
+    0x0000001d : "FILE_DEVICE_SOUND",
+    0x0000001e : "FILE_DEVICE_STREAMS",
+    0x0000001f : "FILE_DEVICE_TAPE",
+    0x00000020 : "FILE_DEVICE_TAPE_FILE_SYSTEM",
+    0x00000038 : "FILE_DEVICE_TERMSRV",
+    0x00000021 : "FILE_DEVICE_TRANSPORT",
+    0x00000022 : "FILE_DEVICE_UNKNOWN",
+    0x0000002c : "FILE_DEVICE_VDM",
+    0x00000023 : "FILE_DEVICE_VIDEO",
+    0x00000024 : "FILE_DEVICE_VIRTUAL_DISK",
+    0x00000025 : "FILE_DEVICE_WAVE_IN",
+    0x00000026 : "FILE_DEVICE_WAVE_OUT",
+}
+
+vollog = logging.getLogger(__name__)
+
+class DeviceTree(interfaces.plugins.PluginInterface):
+    """Listing tree based on drivers and attached devices in a particular windows memory image."""
+
+    _required_framework_version = (2, 0, 1)
+    _version = (1, 0, 0)
+
+    @classmethod
+    def get_requirements(cls) -> List[interfaces.configuration.RequirementInterface]:
+        return [
+            requirements.ModuleRequirement(name = "kernel", description = "Windows kernel",
+                                           architectures = ["Intel32", "Intel64"]),
+            requirements.PluginRequirement(name = "driverscan", plugin = driverscan.DriverScan, version = (1, 0, 0)),
+        ]
+
+    def _generator(self) -> Iterator[Tuple]:
+        kernel = self.context.modules[self.config["kernel"]]
+
+        # Scan the Layer for drivers
+        for driver in driverscan.DriverScan.scan_drivers(self.context, kernel.layer_name, kernel.symbol_table_name):
+            try:
+                driver_name = driver.get_driver_name()
+
+                yield (0, (
+                    format_hints.Hex(driver.vol.offset),
+                    "DRV",
+                    driver_name,
+                    renderers.NotApplicableValue(),
+                    renderers.NotApplicableValue()
+                ))
+
+                # Scan to get the device information of driver.
+                for device in driver.get_devices():
+                    device_name = device.get_device_name()
+                    device_type = DEVICE_CODES.get(device.DeviceType, "UNKNOWN")
```

Comment:
This is ok because we really do want to say that the value is unknown, not that we couldn't get it for some reason.